File: lua/buffer_router.lua
-- BufferRouter.lua
-- A module to manage buffer routing in coc-vue

local api = vim.api

---@class BufferRouter
local BufferRouter = {}

-- Private members
_G.buffers = {}
local current_buffer_id = nil

-- Utility to parse JSON query string
local function parse_query_json(query_json_string)
  if not query_json_string or query_json_string == "" or query_json_string == "{}" then
    return {}
  end
  -- Added type check to prevent error if a table is accidentally passed
  if type(query_json_string) ~= "string" then
    vim.notify("[BufferRouter.lua] parse_query_json was called with a non-string argument: " .. type(query_json_string), vim.log.levels.WARN)
    return {} 
  end
  
  local success, decoded_query = pcall(vim.fn.json_decode, query_json_string)
  if not success then
    vim.notify("[BufferRouter.lua] Failed to decode query JSON: " .. query_json_string, vim.log.levels.ERROR)
    return {}
  end
  return decoded_query
end

-- Utility to parse query string (legacy, might be removed or adapted if still needed elsewhere)
local function parse_query(query_string)
  if not query_string or query_string == "" then
    return {}
  end
  
  local query = {}
  for pair in query_string:gmatch("([^&]+)") do
    local key, value = pair:match("([^=]+)=(.*)")
    if key and value then
      query[key] = value
    end
  end
  
  return query
end

-- Create a new buffer with a path and query
---@param path string The path for the buffer
---@param query_params table|string|nil The query parameters. Expected as a Lua table from TypeScript.
--                                   Can be a URL-like query string from legacy Lua callers (like tests).
---@return string buffer_id The ID of the created buffer
function BufferRouter:create_buffer(path, query_params)
  assert(path, "Path is required to create a buffer")
  
  local id = tostring(math.random(100000, 999999))
  while _G.buffers[id] do
    id = tostring(math.random(100000, 999999))
  end
  
  local final_query_table = {}
  local raw_legacy_query_string = nil -- To store original string if input was string

  if type(query_params) == "table" then
    final_query_table = query_params
  elseif type(query_params) == "string" then
    -- This branch handles legacy callers like test_create_buffer
    final_query_table = parse_query(query_params) -- Use the legacy URL-like string parser
    raw_legacy_query_string = query_params 
  elseif query_params == nil then
    final_query_table = {} -- Default to empty table if nil
  else
    vim.notify("[BufferRouter.lua] create_buffer: query_params is of unhandled type: " .. type(query_params), vim.log.levels.WARN)
    final_query_table = {}
  end
  
  local buf = api.nvim_create_buf(false, true)
  
  _G.buffers[id] = {
    id = id,
    path = path,
    query = final_query_table, -- Consistently a Lua table
    raw_query_string_legacy = raw_legacy_query_string, -- For test assertion or legacy needs
    buf_handle = buf,
    created_at = os.time()
  }
  
  local buffer_name = path
  if final_query_table and next(final_query_table) ~= nil then
    buffer_name = buffer_name .. " [q]" 
  end
  api.nvim_buf_set_name(buf, buffer_name)
  
  -- Return a table with both id and nvimBufferId for better TS integration
  return { id = id, nvimBufferId = buf }
end

-- Get buffer information by ID
---@param id string The ID of the buffer to get information for
---@return table|nil buffer_info Information about the buffer, or nil if not found
function BufferRouter:get_buffer_info(id)
  if not id then
    return nil
  end
  
  local buffer = _G.buffers[id]
  if not buffer then
    return nil
  end
  
  return {
    id = buffer.id,
    path = buffer.path,
    query = buffer.query,
    nvimBufferId = buffer.buf_handle,
    createdAt = buffer.created_at
  }
end

-- Delete a buffer by ID
---@param id string The ID of the buffer to delete
---@return boolean success Whether the deletion was successful
function BufferRouter:delete_buffer(id)
  if not id or not _G.buffers[id] then
    return false
  end
  
  local buffer_info = _G.buffers[id]
  
  -- Delete the buffer from Neovim
  pcall(api.nvim_buf_delete, buffer_info.buf_handle, { force = true })
  
  -- Remove from our tracking
  _G.buffers[id] = nil
  
  -- Reset current buffer if we deleted the current one
  if current_buffer_id == id then
    current_buffer_id = nil
  end
  
  return true
end

-- Switch to a buffer by ID or path
---@param identifier string Either a buffer ID or a path
---@return boolean success Whether the switch was successful
function BufferRouter:switch_buffer(identifier)
  if not identifier then
    return false
  end
  
  local buffer_info = nil
  
  -- Try to find by ID first
  if _G.buffers[identifier] then
    buffer_info = _G.buffers[identifier]
  else
    -- Try to find by path
    for _, buf in pairs(_G.buffers) do
      -- Check if path matches exactly, or if path+query matches
      local full_path = buf.path
      if buf.query_string and buf.query_string ~= "" then
        full_path = full_path .. "?" .. buf.query_string
      end
      
      -- Matching by full_path might be complex if query_json_string is used for naming
      -- For now, primarily match by buf.path or expect ID for switching with complex queries.
      if buf.path == identifier then 
        buffer_info = buf
        break
      end
    end
  end
  
  -- If we didn't find the buffer, return false
  if not buffer_info then
    return false
  end
  
  -- Switch to this buffer
  pcall(api.nvim_set_current_buf, buffer_info.buf_handle)
  current_buffer_id = buffer_info.id
  
  return true
end

-- Get the current buffer information
---@return table|nil buffer_info Information about the current buffer or nil
function BufferRouter:get_current_buffer()
  if not current_buffer_id or not _G.buffers[current_buffer_id] then
    return nil
  end
  
  -- Return a copy of the buffer info (without the handle)
  local buffer_info = _G.buffers[current_buffer_id]
  return {
    id = buffer_info.id,
    path = buffer_info.path,
    query = buffer_info.query,
    query_string = buffer_info.query_string,
    created_at = buffer_info.created_at
  }
end

-- List all managed buffers
---@return table[] buffers Array of buffer information
function BufferRouter:list_buffers()
  local result = {}
  for id, buffer in pairs(_G.buffers) do
    table.insert(result, {
      id = buffer.id,
      path = buffer.path,
      query_string = buffer.query_string,
      created_at = buffer.created_at
    })
  end
  
  return result
end

-- TEST FUNCTIONS
-- These functions are for testing purposes and would typically be in a separate file

-- Test create_buffer
function BufferRouter:test_create_buffer()
  local path = "/test/path"
  local query = "foo=bar&baz=qux"
  
  local id = self:create_buffer(path, query)
  
  assert(id, "Buffer ID should be returned")
  assert(_G.buffers[id], "Buffer should be stored")
  assert(_G.buffers[id].path == path, "Path should match")
  assert(_G.buffers[id].raw_query_string_legacy == query, "Raw legacy query string should match")
  assert(_G.buffers[id].query.foo == "bar", "Query parameter 'foo' should be parsed")
  assert(_G.buffers[id].query.baz == "qux", "Query parameter 'baz' should be parsed")
  
  -- Test with nil query
  local id2 = self:create_buffer("/test/path2", nil)
  assert(id2, "Buffer ID should be returned with nil query")
  assert(_G.buffers[id2], "Buffer should be stored with nil query")
  assert(_G.buffers[id2].path == "/test/path2", "Path should match with nil query")
  assert(_G.buffers[id2].query == {}, "Query should be empty table with nil query")
  
  -- Clean up
  self:delete_buffer(id)
  self:delete_buffer(id2)
  
  print("✓ test_create_buffer passed")
  return true
end

-- Test delete_buffer
function BufferRouter:test_delete_buffer()
  local id = self:create_buffer("/test/path")
  
  assert(_G.buffers[id], "Buffer should exist before deletion")
  
  local success = self:delete_buffer(id)
  
  assert(success, "Deletion should be successful")
  assert(not _G.buffers[id], "Buffer should not exist after deletion")
  
  print("✓ test_delete_buffer passed")
  return true
end

-- Test switch_buffer
function BufferRouter:test_switch_buffer()
  local id1 = self:create_buffer("/test/path1")
  local id2 = self:create_buffer("/test/path2", "param=value")
  
  -- Test switching by ID
  local success1 = self:switch_buffer(id1)
  assert(success1, "Switch by ID should succeed")
  assert(current_buffer_id == id1, "Current buffer should be updated")
  
  -- Test switching by path
  local success2 = self:switch_buffer("/test/path2")
  assert(success2, "Switch by path should succeed")
  assert(current_buffer_id == id2, "Current buffer should be updated")
  
  -- Test switching by path+query
  local success3 = self:switch_buffer("/test/path2?param=value")
  assert(success3, "Switch by path+query should succeed")
  assert(current_buffer_id == id2, "Current buffer should be updated")
  
  -- Clean up
  self:delete_buffer(id1)
  self:delete_buffer(id2)
  
  print("✓ test_switch_buffer passed")
  return true
end

-- Test get_current_buffer
function BufferRouter:test_get_current_buffer()
  local id = self:create_buffer("/test/path", "foo=bar")
  
  -- Switch to the buffer
  self:switch_buffer(id)
  
  local current = self:get_current_buffer()
  
  assert(current, "Current buffer info should be returned")
  assert(current.id == id, "Current buffer ID should match")
  assert(current.path == "/test/path", "Current buffer path should match")
  assert(current.query_string == "foo=bar", "Current buffer query string should match")
  assert(current.query.foo == "bar", "Current buffer query parameter should match")
  
  -- Clean up
  self:delete_buffer(id)
  
  print("✓ test_get_current_buffer passed")
  return true
end

-- Test list_buffers
function BufferRouter:test_list_buffers()
  -- Clear existing buffers
  for id, _ in pairs(_G.buffers) do
    self:delete_buffer(id)
  end
  
  -- Create test buffers
  local id1 = self:create_buffer("/test/list/path1")
  local id2 = self:create_buffer("/test/list/path2", "param=value")
  
  -- Get buffer list
  local buffers = self:list_buffers()
  
  assert(#buffers == 2, "Should list 2 buffers")
  
  -- Verify first buffer
  local found_id1 = false
  local found_id2 = false
  
  for _, buf in ipairs(buffers) do
    if buf.id == id1 then
      found_id1 = true
      assert(buf.path == "/test/list/path1", "Buffer 1 path should match")
    elseif buf.id == id2 then
      found_id2 = true
      assert(buf.path == "/test/list/path2", "Buffer 2 path should match")
      assert(buf.query_string == "param=value", "Buffer 2 query string should match")
    end
  end
  
  assert(found_id1, "Buffer 1 should be in the list")
  assert(found_id2, "Buffer 2 should be in the list")
  
  -- Clean up
  self:delete_buffer(id1)
  self:delete_buffer(id2)
  
  print("✓ test_list_buffers passed")
  return true
end

-- Run all tests
function BufferRouter:run_tests()
  print("Running BufferRouter tests...")
  self:test_create_buffer()
  self:test_delete_buffer()
  self:test_switch_buffer()
  self:test_get_current_buffer()
  self:test_list_buffers()
  print("All BufferRouter tests passed!")
end

return BufferRouter

File: src/bufferRouter.ts
import { ExtensionContext, workspace, Disposable } from 'coc.nvim';
import { EventEmitter } from 'events';
import { bridgeCore, BridgeMessage } from './bridge/core'; // Corrected path, removed MessageType

// Import events differently to avoid undefined issues in tests
let events: any;
try {
  // This import may fail in test environments
  events = require('coc.nvim').events;
} catch (e) {
  // Create mock events in test environment
  events = {
    on: () => ({ dispose: () => {} })
  };
}

/**
 * Interface representing a buffer route
 */
export interface BufferRoute {
  id: string;
  path: string;
  query: Record<string, any>;
  createdAt: number;
  nvimBufferId?: number; // Optional Neovim buffer ID for easier validation
}

/**
 * Type for buffer content updates
 * - undefined: keep the line unchanged
 * - null: delete the line
 * - string: add or update the line
 */
export type BufferUpdate = (string | null | undefined)[];

/**
 * Class that bridges Lua buffer router with Coc.nvim
 * Provides methods to create, delete, and switch buffers with full reactivity support.
 * Implements an EventEmitter-based reactive system to notify subscribers of buffer changes.
 * Handles errors gracefully to ensure UI consistency and prevent crashes.
 * Core component for buffer management that integrates TypeScript with Lua-based Neovim buffers.
 */
export class BufferRouter implements Disposable {
  private callbackRegistry: Map<string, (...args: any[]) => void> = new Map();
  private nextCallbackId = 0;

  private nvim = workspace.nvim;
  private disposables: Disposable[] = [];
  private emitter = new EventEmitter();
  private currentBuffer: BufferRoute | null = null;
  
  /**
   * Event names used by the BufferRouter
   */
  static readonly Events = {
    BUFFER_CREATED: 'buffer-created',
    BUFFER_DELETED: 'buffer-deleted',
    BUFFER_SWITCHED: 'buffer-switched',
    BUFFER_UPDATED: 'buffer-updated',
    CURRENT_BUFFER_CHANGED: 'current-buffer-changed'
  };
  
  constructor(context: ExtensionContext) {
    // Increase max listeners to prevent memory leak warnings
    this.emitter.setMaxListeners(20);
    
    // Ensure Lua buffer_router is loaded when this class is instantiated
    this.loadLuaModule();
    
    // Listen to buffer events from Neovim
    this.setupEventListeners(context);
    
    // Initialize the current buffer
    this.refreshCurrentBuffer();

    // Register a handler with bridgeCore to execute callbacks from Lua
    bridgeCore.registerHandler('execute_buffer_callback', (message: BridgeMessage) => {
      if (message.payload && typeof message.payload.callbackId === 'string' && Array.isArray(message.payload.args)) {
        this.executeCallback(message.payload.callbackId, message.payload.args);
      } else {
        console.error('[BufferRouter] Invalid payload for execute_buffer_callback:', message.payload);
      }
    });
  }
  
  /**
   * Set up event listeners for buffer changes
   * @private
   */
  private setupEventListeners(context: ExtensionContext): void {
    try {
      // Only set up event listeners if events is properly defined
      if (events && typeof events.on === 'function') {
        // Listen to buffer events from Neovim
        const bufferChangedDisposable = events.on(
          'BufEnter',
          async () => {
            await this.refreshCurrentBuffer();
          },
          null,
          context.subscriptions
        );
        
        this.disposables.push(bufferChangedDisposable);
      } else {
        console.log('[BufferRouter] Events API not available, using polling for buffer changes');
        // Set up polling as fallback
        const intervalId = setInterval(() => {
          this.refreshCurrentBuffer().catch(e => {
            console.error('Error refreshing buffer in polling mode:', e);
          });
        }, 1000);
        
        // Create a disposable for the interval
        this.disposables.push({
          dispose: () => clearInterval(intervalId)
        });
      }
    } catch (error) {
      console.error('[BufferRouter] Error setting up event listeners:', error);
    }
  }
  
  /**
   * Refresh the current buffer information
   * Public for testing purposes
   */
  public async refreshCurrentBuffer(): Promise<void> {
    try {
      const buffer = await this.getCurrentBuffer();
      
      // Check if the buffer has changed
      const hasChanged = this.hasBufferChanged(this.currentBuffer, buffer);
      
      if (hasChanged) {
        const oldBuffer = this.currentBuffer;
        this.currentBuffer = buffer;
        
        // Emit the change event with old and new buffer
        this.emitter.emit(BufferRouter.Events.CURRENT_BUFFER_CHANGED, {
          oldBuffer,
          newBuffer: buffer
        });
      }
    } catch (error) {
      console.error('Error refreshing current buffer:', error);
    }
  }
  
  /**
   * Check if a buffer has changed
   * @private
   * @param oldBuffer - Previous buffer state
   * @param newBuffer - New buffer state
   * @returns Whether the buffer has changed
   */
  private hasBufferChanged(oldBuffer: BufferRoute | null, newBuffer: BufferRoute | null): boolean {
    // If both are null, no change
    if (oldBuffer === null && newBuffer === null) return false;
    
    // If one is null and the other isn't, that's a change
    if (oldBuffer === null || newBuffer === null) return true;
    
    // Compare relevant properties
    return (
      oldBuffer.id !== newBuffer.id ||
      oldBuffer.path !== newBuffer.path ||
      JSON.stringify(oldBuffer.query) !== JSON.stringify(newBuffer.query)
    );
  }
  
  /**
   * Load the Lua buffer_router module
   * @private
   */
  private async loadLuaModule(): Promise<void> {
    await this.nvim.command('lua require("buffer_router")')
  }
  
  /**
   * Call a method on the Lua BufferRouter
   * @private
   * @param method - Method name to call
   * @param args - Arguments to pass to the method
   * @returns Result from the Lua method
   */
  private async callLuaMethod<T>(method: string, ...args: any[]): Promise<T> {
    const serializeArgForLua = (arg: any): string => {
      if (typeof arg === 'string') {
        // Escape single quotes and backslashes for Lua string literals
        const escaped = arg.replace(/'/g, "\\'").replace(/\\/g, '\\\\');
        return `'${escaped}'`;
      }
      if (arg === null || arg === undefined) {
        return 'nil';
      }
      if (typeof arg === 'boolean') {
        return arg ? 'true' : 'false';
      }
      if (typeof arg === 'number') {
        return String(arg);
      }
      if (Array.isArray(arg)) {
        return `{${arg.map(serializeArgForLua).join(', ')}}`;
      }
      if (typeof arg === 'object') {
        const L: string[] = [];
        for (const key in arg) {
          if (Object.prototype.hasOwnProperty.call(arg, key)) {
            // Assuming keys are simple identifiers for Lua.
            // If keys could be complex (e.g., contain spaces or special chars),
            // they would need to be quoted: `['${key}'] = ${serializeArgForLua(arg[key])}`
            L.push(`${key} = ${serializeArgForLua(arg[key])}`);
          }
        }
        return `{${L.join(', ')}}`;
      }
      // Fallback for unknown types
      console.warn(`[BufferRouter] Unknown type for Lua serialization: ${typeof arg}`, arg);
      const escapedFallback = String(arg).replace(/'/g, "\\'").replace(/\\/g, '\\\\');
      return `'${escapedFallback}'`;
    };

    const luaArgs = args.map(serializeArgForLua).join(', ');
    const luaCommand = `return require('buffer_router'):${method}(${luaArgs})`;

    try {
      return await this.nvim.lua(luaCommand) as T;
    } catch (error) {
      console.error(`[BufferRouter] Error executing Lua command: ${luaCommand}`, error);
      throw error; // Re-throw to be caught by the caller
    }
  }

  /**
   * Register a callback function and return its ID.
   * @private
   * @param func - The callback function to register.
   * @returns The ID of the registered callback.
   */
  private registerCallback(func: (...args: any[]) => void): string {
    const callbackId = `cb_${this.nextCallbackId++}`;
    this.callbackRegistry.set(callbackId, func);
    return callbackId;
  }

  /**
   * Execute a registered callback by its ID.
   * This method is called by bridgeCore when Lua requests callback execution.
   * @param callbackId - The ID of the callback to execute.
   * @param args - Arguments to pass to the callback function.
   */
  public executeCallback(callbackId: string, args: any[]): void {
    const func = this.callbackRegistry.get(callbackId);
    if (func) {
      try {
        func(...args);
      } catch (e) {
        console.error(`[BufferRouter] Error executing callback ${callbackId}:`, e);
      }
    } else {
      console.warn(`[BufferRouter] Callback with ID ${callbackId} not found.`);
    }
  }

  public async createBuffer(path: string, query?: Record<string, any>): Promise<string | null> {
    try {
      // Convert query object to query string if provided
      const processedQuery: Record<string, any> = {};
      if (query) {
        for (const key in query) {
          if (Object.prototype.hasOwnProperty.call(query, key)) {
            if (typeof query[key] === 'function' && key.startsWith('on') && key.length > 2 && key[2] === key[2].toUpperCase()) {
              const callbackId = this.registerCallback(query[key] as (...args: any[]) => void);
              processedQuery[key] = { __isCallback: true, id: callbackId };
            } else {
              // Other props are passed as is. Functions not matching on<Event> convention
              // will likely become null or be omitted by JSON.stringify in callLuaMethod.
              processedQuery[key] = query[key];
            }
          }
        }
      }

      const bufferInfo = await this.callLuaMethod<{id: string, nvimBufferId?: number}>('create_buffer', path, processedQuery);
      if (bufferInfo && bufferInfo.id) {
        const bufferId = bufferInfo.id;
        // Get the full buffer data to emit in the event
        const bufferData: BufferRoute = {
          id: bufferId,
          path,
          query: processedQuery, // Use processedQuery here
          createdAt: Date.now(),
          nvimBufferId: bufferInfo.nvimBufferId
        };
        
        // Emit event for buffer creation
        this.emitter.emit(BufferRouter.Events.BUFFER_CREATED, bufferData);
        
        // Refresh current buffer after creation
        await this.refreshCurrentBuffer();
        
        return bufferId;
      }
      return null;
    } catch (error) {
      console.error('Error creating buffer:', error);
      return null;
    }
  }
  
  /**
   * Delete a buffer by ID
   * @param id - Buffer ID to delete
   * @returns Success status, false if deletion failed
   */
  public async deleteBuffer(id: string): Promise<boolean> {
    try {
      // Get buffer info before deletion if possible
      let bufferInfo: BufferRoute | null = null;
      try {
        // Try to get the buffer info if it's the current one
        const current = await this.getCurrentBuffer();
        if (current && current.id === id) {
          bufferInfo = current;
        }
      } catch (e) {
        // Ignore errors when trying to get buffer info
      }
      
      const success = await this.callLuaMethod<boolean>('delete_buffer', id);
      
      if (success) {
        // Emit event for buffer deletion
        this.emitter.emit(BufferRouter.Events.BUFFER_DELETED, {
          id,
          bufferInfo
        });
        
        // Refresh current buffer after deletion
        await this.refreshCurrentBuffer();
      }
      
      return success;
    } catch (error) {
      console.error('Error deleting buffer:', error);
      return false;
    }
  }
  
  /**
   * Switch to a buffer by ID or path
   * @param identifier - Buffer ID or path
   * @returns Success status
   */
  public async switchBuffer(identifier: string): Promise<boolean> {
    try {
      const success = await this.callLuaMethod<boolean>('switch_buffer', identifier);
      
      if (success) {
        // Emit event for buffer switch
        this.emitter.emit(BufferRouter.Events.BUFFER_SWITCHED, { identifier });
        
        // Get and emit current buffer info after switch
        await this.refreshCurrentBuffer();
      }
      
      return success;
    } catch (error) {
      console.error('Error switching buffer:', error);
      return false;
    }
  }
  
  /**
   * Get current buffer information
   * @returns Current buffer route or null
   */
  public async getCurrentBuffer(): Promise<BufferRoute | null> {
    try {
      return await this.callLuaMethod<BufferRoute | null>('get_current_buffer');
    } catch (error) {
      console.error('Error refreshing current buffer:', error);
      return null;
    }
  }
  
  /**
   * Get information about a specific buffer by ID
   * @param bufferId - The ID of the buffer to get information for
   * @returns Buffer route information or null if not found
   */
  public async getBufferInfo(bufferId: string): Promise<BufferRoute | null> {
    try {
      // Get buffer information from Lua
      const info = await this.callLuaMethod<BufferRoute | null>('get_buffer_info', bufferId);
      
      if (!info) {
        console.warn(`[BufferRouter] No information available for buffer ${bufferId}`);
        return null;
      }
      
      return info;
    } catch (error) {
      console.error(`[BufferRouter] Error getting buffer info for ${bufferId}:`, error);
      return null;
    }
  }

  /**
   * Update the content of a buffer with new lines
   * @param bufferId - Buffer ID
   * @param lines - Array of operations (undefined = keep, null = delete, string = insert/replace)
   * @returns Whether the update was successful
   */
  public async updateBufferContent(bufferId: number, lines: BufferUpdate): Promise<boolean> {
    try {
      // Call the Lua utils render_buffer_diff function to apply the diff
      await this.nvim.lua(`return require('vue-ui.utils.render_buffer').ApplyBufferDiff(${bufferId}, vim.fn.json_decode('${JSON.stringify(lines)}'))`);
      
      // Emit event for buffer update
      this.emitter.emit(BufferRouter.Events.BUFFER_UPDATED, { 
        bufferId, 
        linesUpdated: lines.filter(line => line !== undefined).length 
      });
      
      return true;
    } catch (error) {
      console.error(`Error updating buffer content: ${error}`);
      return false;
    }
  }
  
  /**
   * Subscribe to buffer events
   * @param event - Event name to subscribe to
   * @param listener - Callback function for the event
   * @returns Dispose function to unsubscribe
   */
  public on(event: string, listener: (...args: any[]) => void): Disposable {
    this.emitter.on(event, listener);
    
    return {
      dispose: () => {
        this.emitter.removeListener(event, listener);
      }
    };
  }
  
  /**
   * Get current buffer without triggering reactive updates
   * @returns Current buffer or null if none
   */
  public getCurrentBufferSync(): BufferRoute | null {
    return this.currentBuffer;
  }
  
  /**
   * Dispose of resources
   */
  public dispose(): void {
    // Remove all event listeners
    this.emitter.removeAllListeners();
    
    // Dispose of all disposables
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }
}

/**
 * Hook for using buffer router in components
 * @param context - Extension context
 * @returns BufferRouter API
 */
export function useBufferRouter(context: ExtensionContext) {
  const router = new BufferRouter(context);
  
  return {
    // Buffer operations
    createBuffer: router.createBuffer.bind(router),
    deleteBuffer: router.deleteBuffer.bind(router),
    switchBuffer: router.switchBuffer.bind(router),
    getCurrentBuffer: router.getCurrentBuffer.bind(router),
    getCurrentBufferSync: router.getCurrentBufferSync.bind(router),
    
    // Event handling
    on: router.on.bind(router),
    events: BufferRouter.Events,
    
    // Cleanup
    dispose: router.dispose.bind(router)
  };
}

// Test utilities - these would typically be in a separate test file

/**
 * Run tests for the BufferRouter
 */
export async function testBufferRouter(context: ExtensionContext): Promise<void> {
  const router = new BufferRouter(context);
  
  // Test createBuffer
  const testCreateBuffer = async () => {
    const path = '/test/typescript/path';
    const query = { foo: 'bar', baz: 'qux' };
    
    const id = await router.createBuffer(path, query);
    console.assert(id, 'Buffer ID should be returned');
    
    if (!id) {
      console.error('Failed to create buffer for testing');
      return false;
    }
    
    // Verify current buffer can be retrieved
    const buffer = await router.getCurrentBuffer();
    console.assert(buffer !== null, 'Current buffer should be retrievable');
    
    await router.deleteBuffer(id);
    
    return true;
  };
  
  // Test deleteBuffer
  const testDeleteBuffer = async () => {
    const id = await router.createBuffer('/test/typescript/delete');
    if (!id) {
      console.error('Failed to create buffer for delete test');
      return false;
    }
    
    const result = await router.deleteBuffer(id);
    console.assert(result, 'Deletion should be successful');
    
    return true;
  };
  
  // Test switchBuffer
  const testSwitchBuffer = async () => {
    const id1 = await router.createBuffer('/test/typescript/switch1');
    if (!id1) {
      console.error('Failed to create first buffer for switch test');
      return false;
    }
    
    const id2 = await router.createBuffer('/test/typescript/switch2', { param: 'value' });
    if (!id2) {
      console.error('Failed to create second buffer for switch test');
      await router.deleteBuffer(id1);
      return false;
    }
    
    // Test switching by ID
    let result = await router.switchBuffer(id1);
    console.assert(result, 'Switch by ID should succeed');
    
    // Test switching by path
    result = await router.switchBuffer('/test/typescript/switch2');
    console.assert(result, 'Switch by path should succeed');
    
    // Clean up
    await router.deleteBuffer(id1);
    await router.deleteBuffer(id2);
    
    return true;
  };
  
  // Test getCurrentBuffer
  const testGetCurrentBuffer = async () => {
    const path = '/test/typescript/current';
    const query = { test: 'value' };
    
    const id = await router.createBuffer(path, query);
    if (!id) {
      console.error('Failed to create buffer for getCurrentBuffer test');
      return false;
    }
    
    const switchResult = await router.switchBuffer(id);
    if (!switchResult) {
      console.error('Failed to switch to buffer for getCurrentBuffer test');
      await router.deleteBuffer(id);
      return false;
    }
    
    const current = await router.getCurrentBuffer();
    console.assert(current, 'Current buffer info should be returned');
    console.assert(current?.id === id, 'Current buffer ID should match');
    console.assert(current?.path === path, 'Current buffer path should match');
    console.assert(current?.query.test === 'value', 'Current buffer query should match');
    
    await router.deleteBuffer(id);
    
    return true;
  };
  
  // Run all tests
  try {
    await testCreateBuffer();
    await testDeleteBuffer();
    await testSwitchBuffer();
    await testGetCurrentBuffer();
    console.log('All BufferRouter TypeScript tests passed!');
  } catch (error) {
    console.error('BufferRouter tests failed:', error);
  } finally {
    router.dispose();
  }
}
